### DISCUSSION TITLE
Using fzf for tagging workflow - ability to add to input options?

### DISCUSSION THREAD

**Discussion by sullyj3** [junegunn/fzf#3662] [Q&A] [answered]
I'm trying to use fzf for tagging files. I have a shell script which opens files in a loop, presenting you with a fzf window for selecting tags for each one. I'm using multi-select mode, because obviously a file can have multiple tags.

The complicated part is that I want to be able to tag with both pre-existing tags that were piped into fzf, as well as tags that don't yet exist, and therefore aren't in the input. In other words, I'm essentially using fzf as a fancy auto-complete, to make adding existing long tags less annoying to type.

I found #3232 which mentions the trick of using `--print-query` to output the current query in addition to the selected options, but this isn't a good fit for multi-selections, because
- you can't add multiple custom tags
- as soon as you add one custom tag, you've exited fzf, so you can't even add any more non-custom tags.

Does anyone know of a way to support this use-case of multi-selecting from both an existing set, and additional new options supplied by the user?

**Comment by junegunn** [comment, marked as answer]
This should give you some ideas.

```sh
cat /usr/share/dict/words | fzf --header 'Press space to add the typed word' --multi \
  --bind 'start:execute-silent(: > /tmp/added)' \
  --bind 'space:execute-silent(echo {q} >> /tmp/added)+clear-query+refresh-preview' \
  --preview '
    echo "Selected words:"
    echo {+} | sed "s/^/  - /"
    echo
    echo "Added words:"
    cat /tmp/added | sed "s/^/  - /"
  '
```

**Reply by sullyj3** [reply]
Thanks very much, I'll see if I can figure this out.

**Reply by lamyergeier** [reply]
@sullyj3 Could you please share the shell script. I am also interesting in this idea and would like to explore more on it. 

PS: I have modified the above code snippet to also print the added tags, and I use @ instead of space as I also use phrase as a tag

```bash
  mapfile -t TagA < <( command cat Data/Tag.txt | fzf --header 'Press @ to add the typed word' --multi \
    --bind 'start:execute-silent(: > /tmp/added)' \
    --bind '@:execute-silent(echo {q} >> /tmp/added)+clear-query+refresh-preview' \
    --preview '
      echo "Selected words:"
      echo {+} | sed "s/^/  - /"
      echo
      echo "Added words:"
      cat /tmp/added | sed "s/^/  - /"
    ' && command cat /tmp/added)
```

**Comment by lamyergeier** [comment]
@junegunn From your following code snippet, in certain cases when none of the words from the file is suitable, is it possible to not choose any output from `/usr/share/dict/words` and choose only the newly added word(s)?

> This should give you some ideas.
> 
> ```shell
> cat /usr/share/dict/words | fzf --header 'Press space to add the typed word' --multi \
>   --bind 'start:execute-silent(: > /tmp/added)' \
>   --bind 'space:execute-silent(echo {q} >> /tmp/added)+clear-query+refresh-preview' \
>   --preview '
>     echo "Selected words:"
>     echo {+} | sed "s/^/  - /"
>     echo
>     echo "Added words:"
>     cat /tmp/added | sed "s/^/  - /"
>   '
> ```

May be we could bind '#' to just choose the newly added words. Is it possible? Or is there any better solution?

**Reply by junegunn** [reply]
You might want to check out `transform` action to conditionally perform different sets of actions using a shell script.
